utils/src/acknowledgement.rs

```diff
@@ -0,0 +1,160 @@
+//! Utilities for providing acknowledgement.
+
+use core::fmt::Debug;
+use futures::channel::oneshot;
+use std::{
+    future::Future,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Arc, Mutex,
+    },
+};
+
+/// A mechanism for acknowledging the completion of a task.
+pub trait Acknowledgement: Clone + Send + Sync + Debug + 'static {
+    /// Future resolved once the acknowledgement is handled.
+    type Waiter: Future<Output = Result<(), Self::Error>> + Send + Sync + Unpin + 'static;
+    /// Error produced if the acknowledgement is not handled.
+    type Error: Debug + Send + Sync + 'static;
+
+    /// Create a new acknowledgement handle paired with the waiter.
+    fn handle() -> (Self, Self::Waiter);
+
+    /// Fulfill the acknowledgement.
+    fn acknowledge(self);
+}
+
+/// Default acknowledgement implementation that requires a minimum number of acknowledgements.
+///
+/// If any acknowledgement is not handled (before the minimum number of acknowledgements is received), the acknowledgement will be cancelled.
+pub struct Min<const N: usize = 1> {
+    state: Arc<AckState<N>>,
+    acknowledged: bool,
+}
+
+impl<const N: usize> Clone for Min<N> {
+    fn clone(&self) -> Self {
+        Self {
+            state: self.state.clone(),
+            acknowledged: false,
+        }
+    }
+}
+
+impl<const N: usize> std::fmt::Debug for Min<N> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Min").field("listeners", &N).finish()
+    }
+}
+
+impl<const N: usize> Drop for Min<N> {
+    fn drop(&mut self) {
+        if self.acknowledged {
+            return;
+        }
+
+        // If not yet acknowledged, cancel the acknowledgement.
+        self.state.cancel();
+        self.acknowledged = true;
+    }
+}
+
+impl<const N: usize> Acknowledgement for Min<N> {
+    type Error = oneshot::Canceled;
+    type Waiter = oneshot::Receiver<()>;
+
+    fn handle() -> (Self, Self::Waiter) {
+        assert!(N > 0, "requires N > 0 listeners");
+        let (tx, rx) = oneshot::channel();
+        (
+            Self {
+                state: Arc::new(AckState::new(tx)),
+                acknowledged: false,
+            },
+            rx,
+        )
+    }
+
+    fn acknowledge(mut self) {
+        self.state.acknowledge();
+        self.acknowledged = true;
+    }
+}
+
+/// State for the [Min] acknowledgement.
+struct AckState<const N: usize> {
+    sender: Mutex<Option<oneshot::Sender<()>>>,
```

Comment:
Let's use a `Mutex` that can yield here (https://docs.rs/futures/latest/futures/lock/struct.Mutex.html).

Tests pass because the deterministic runtime doesn't have a scheduler, but this will cause issues on the tokio runtime if `acknowledge` is called within a future that's currently in the same thread as the runtime's scheduler.

`AckState::acknowledge` & `Acknowledgement::acknowledge` should also be colored async to facilitate this.

Reply:
oneshot sender isn't async, so I didn't feel it was necessary?

Reply:
We, unfortunately, need to call `cancel()` in `drop` ... so, adding async locking here would be pretty gnar.

I also believe we really shouldn't see contention on the mutex (it should only be called once unless `cancel()` -> but that should be rare).

Reply:
https://github.com/commonwarexyz/monorepo/pull/2217